source/slang/slang-check-conversion.cpp

```diff
@@ -639,6 +639,16 @@ namespace Slang
             return true;
         }
 
+        // If both are string types we assume they are convertable in both directions
+        if (as<StringTypeBase>(fromType) && as<StringTypeBase>(toType))
+        {
+            if (outToExpr)
+                *outToExpr = fromExpr;
+            if (outCost)
+                *outCost = kConversionCost_None;
+            return true;
+        }
```

Comment:
Have zero-cost conversion in both directions seems like it would make it impossible for the compiler to pick between overloads that take the two different string types:

```
void f(String s);
void f(NativeString s);

...
NativeString n = "...";
f(n); // which one does it call?
```

We probably need to define one of the directions as being a "widening" conversion (probably from `NativeString` to `String`) and the other as a "narrowing" conversion (probably from `String` to `NativeString`). That would allow both to be implicit but not zero-cost.

Another note here is that we *should* be able to implement the two conversion directions this implies by adding appropriately annotated `__init` declarations to the stdlib types, rather than making them be built-in compiler magic.